tools/mockgen/mockgen.go

```diff
@@ -0,0 +1,102 @@
+// Package mockgen contains a wrapper for the `mockgen` tool, which bypasses running `mockgen` if the destination file
+// is newer than the source file. This is useful because `mockgen` is slow and can be a bottleneck in the build process.
+package mockgen
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"syscall"
+)
+
+// Run a cached version of `mockgen`, which checks the modification times of the source and destination files,
+// running mockgen only if necessary. This is similar to the behavior of `Make`, but it wasn't easy to express this
+// in a Makefile because the generate commands are in our source code.
+func Run(args []string, opts ...Option) error {
+	params := &Params{
+		Exec: runRealCommand,
+	}
+	for _, opt := range opts {
+		opt(params)
+	}
+
+	upToDate, err := isDestinationFileUpToDateWithSourceInArgs(args)
+	if err != nil {
+		return err
+	}
+	if !upToDate {
+		return nil
+	}
+
+	// We couldn't confirm that the target file was up-to-date; run mockgen
+	return params.Exec(args)
+}
+
+// Params for running the mockgen wrapper.
+type Params struct {
+	// Exec is a function that should execute a shell command with the given arguments.
+	Exec func(args []string) error
```

Comment:
the options pattern feels like overkill for overriding one function with a mock.. maybe you want an `Executor` interface and then you can mock it with mockgen :)

Reply:
Removed optiosn pattern usage but still left an argument so tests can provide their own version.